README.md
# uCRC_t - is C++ class for calculation CRC any sizes 1-64 bits


## Description

`uCRC_t` is universal C++ class for calculation CRC any sizes(width) 1-64 bits.


#### Features of the implementation:

 - The code has no dependencies.
 - You can change the parameters of the algorithm CRC in Run-Time, see `set_xxx` methods.
 - For **any** bit-depth (width) of CRC will be use the standard table method for calculation. Will be calculated standart table for byte (table size 256 elements)


#### Limitations of the implementation:

- This class is universal, but it leads to extra work because the data type is `uint64_t`. This is not optimal.


If you don't need such universality, you can use a more optimized template for your algorithm see:[CRC_t C++ Template](https://github.com/KoynovStas/CRC_t)


Class parameters is the standard Specifications algorithms CRC as described in Ross N. Williams [A PAINLESS GUIDE TO CRC ERROR DETECTION ALGORITHMS](./doc/crc_v3.txt)


**Ross N. Williams:**

> By putting all these pieces together we end up with the parameters of the algorithm:
>
>   - **NAME:** This is a name given to the algorithm. A string value.
>
>   - **WIDTH:** This is the width of the algorithm expressed in bits.
>   This is one less than the width of the Poly.
>
>   - **POLY:** This parameter is the poly. This is a binary value that
>   should be specified as a hexadecimal number. The top bit of the
>   poly should be omitted. For example, if the poly is 10110, you
>   should specify 06. An important aspect of this parameter is that it
>   represents the unreflected poly; the bottom bit of this parameter
>   is always the LSB of the divisor during the division regardless of
>   whether the algorithm being modelled is reflected.
>
>   - **INIT:** This parameter specifies the initial value of the register
>   when the algorithm starts. This is the value that is to be assigned
>   to the register in the direct table algorithm. In the table
>   algorithm, we may think of the register always commencing with the
>   value zero, and this value being XORed into the register after the
>   N'th bit iteration. This parameter should be specified as a
>   hexadecimal number.
>
>   - **REFIN:** This is a boolean parameter. If it is FALSE, input bytes are
>   processed with bit 7 being treated as the most significant bit
>   (MSB) and bit 0 being treated as the least significant bit. If this
>   parameter is FALSE, each byte is reflected before being processed.
>
>   - **REFOUT:** This is a boolean parameter. If it is set to FALSE, the
>   final value in the register is fed into the XOROUT stage directly,
>   otherwise, if this parameter is TRUE, the final register value is
>   reflected first.
>
>   - **XOROUT:** This is an W-bit value that should be specified as a
>   hexadecimal number. It is XORed to the final register value (after
>   the REFOUT) stage before the value is returned as the official
>   checksum.
>
>   - **CHECK:** This field is not strictly part of the definition, and, in
>   the event of an inconsistency between this field and the other
>   field, the other fields take precedence. This field is a check
>   value that can be used as a weak validator of implementations of
>   the algorithm. The field contains the checksum obtained when the
>   ASCII string "123456789" is fed through the specified algorithm
>   (i.e. 313233... (hexadecimal)).




**The class has the following public methods:**
```C++
explicit uCRC_t(std::string Name   = "CRC-32",
                uint8_t     Bits   = 32,
                uint64_t    Poly   = 0x04c11db7,
                uint64_t    Init   = 0xffffffff,
                bool        RefIn  = true,
                bool        RefOut = true,
                uint64_t    XorOut = 0xffffffff) noexcept;


explicit uCRC_t(uint8_t  Bits,
                uint64_t Poly,
                uint64_t Init,
                bool     RefIn,
                bool     RefOut,
                uint64_t XorOut) noexcept;


std::string name;


// get param CRC
uint8_t  get_bits()    const noexcept{ return bits;   }
uint64_t get_poly()    const noexcept{ return poly;   }
uint64_t get_init()    const noexcept{ return init;   }
uint64_t get_xor_out() const noexcept{ return xor_out;}
bool     get_ref_in()  const noexcept{ return ref_in; }
bool     get_ref_out() const noexcept{ return ref_out;}

uint64_t get_crc_init()const noexcept{ return crc_init;} //crc_init = reflect(init, bits) if RefIn, else = init
uint64_t get_top_bit() const noexcept{ return top_bit; }
uint64_t get_crc_mask()const noexcept{ return crc_mask;}
uint64_t get_check()   const noexcept;                  //crc for ASCII string "123456789" (i.e. 313233... (hexadecimal)).


// set param CRC
int  set_bits(uint8_t new_value)     noexcept;
void set_poly(uint64_t new_value)    noexcept{ poly    = new_value; init_class();}
void set_init(uint64_t new_value)    noexcept{ init    = new_value; init_class();}
void set_ref_in(bool new_value)      noexcept{ ref_in  = new_value; init_class();}
void set_ref_out(bool new_value)     noexcept{ ref_out = new_value;}
void set_xor_out(uint64_t new_value) noexcept{ xor_out = new_value;}


// Calculate methods
uint64_t get_crc(const void* data, size_t len) const noexcept;
int      get_crc(uint64_t &crc, const char* file_name) const noexcept;
int      get_crc(uint64_t &crc, const char* file_name, void* buf, size_t size_buf) const noexcept;


// Calculate for chunks of data
uint64_t get_raw_crc(const void* data, size_t len) const noexcept;                   //get raw_crc for first chunk of data
uint64_t get_raw_crc(const void* data, size_t len, uint64_t raw_crc) const noexcept; //get raw_crc for chunk of data
uint64_t get_end_crc(uint64_t raw_crc) const noexcept;
```

More details see: **[ucrc_t.h](./ucrc_t.h)**



## Usage

**To start working, perform the following steps:**

1. You must select the appropriate version:
  + C++98 version - 1.x
  + C++11 version - 2.x
    > Where .x maximum minor version at the moment
2. You need to include **[ucrc_t.h](./ucrc_t.h)** file in your **.cpp** file.
3. And add file **[ucrc_t.cpp](./ucrc_t.cpp)** to list of source files to compile.
4. Set parameters to a class (see an examples).

**Note:**
> The parameters for the class correspond to the `Ross N. Williams` specification.
A list of CRC algorithms for this specification can be found here: [Catalogue CRC algorithms](http://reveng.sourceforge.net/crc-catalogue/all.htm)




## Examples

**Get CRC-32 for file:**

```C++
uint64_t crc;

uCRC_t ucrc;

int err = ucrc.get_crc(crc, "std_file_to_test_crc");

if( !err )
    //uses crc
```


**Note:**
>
>  methods for calculate CRC file
>
>  ```C++
>  int get_crc(uint64_t &crc, const char* file_name) const noexcept;
>  ```
>  These method are reentrant. He use a buffer on the stack.
>  The buffer size is 4 Kib (4096 bytes) which is optimal for most systems.
>  If you have a buffer or needs aligned buffer, you can use the following method:
>
>  ```C++
>  int get_crc(uint64_t &crc, const char* file_name, void* buf, size_t size_buf) const noexcept;
>  ```

**Get CRC-32 for single buf:**

```C++
char buf[len_of_buf]; //bla bla

uint32_t crc;

uCRC_t ucrc(32, 0x04C11DB7, 0xFFFFFFFF, true, true, 0xFFFFFFFF);

crc = ucrc.get_crc(buf, len_of_buf);
//uses crc
```


**Get CRC-ModBus for single buf:**

```C++
char buf[len_of_buf]; //bla bla

uint16_t crc;

uCRC_t ucrc(16, 0x8005, 0xffff, true, true, 0x0);

crc = ucrc.get_crc(buf, len_of_buf);
//uses crc
```


**Get CRC-8 for buf(s) (chunks):**

**Note:**
> when the method is used `uint64_t get_raw_crc(const void* data, size_t len, uint64_t crc)`
for the first byte (or chunk of data) **crc** param must be obtained through a method `get_crc_init()` and in the final you need to call the method: `get_end_crc()`:

```C++
char buf[len_of_buf];   //bla bla
char buf2[len_of_buf2]; //bla bla

uint8_t crc;

uCRC_t ucrc(8, 0x7,  0x0, false, false, 0x0);

crc = ucrc.get_raw_crc(buf,  len_of_buf);        //first chunk (Since from version 1.4)
crc = ucrc.get_raw_crc(buf2, len_of_buf2, crc);  //second chunk
crc = ucrc.get_end_crc(crc);

//uses crc
```

More details can be found in the test application: **[crc_test.cpp](./crc_test.cpp)**



## Tests

To run tests, perform the following steps:

```
cd tests
cmake . -B ./build
cmake --build build
```



## License

[BSD-3-Clause](./LICENSE).


CHANGELOG.md
# Change Log


[**uCRC_t**](https://github.com/KoynovStas/uCRC_t) - is C++ class for calculation CRC any sizes 1-64 bits

License: BSD 3-Clause

Copyright (c) 2015, Koynov Stas - skojnov@yandex.ru



---
## [v2.0](https://github.com/KoynovStas/uCRC_t/tree/v2.0) (2023-05-27)
[Full Changelog](https://github.com/KoynovStas/uCRC_t/compare/v1.4...v2.0)


### New
 - New: Since version 2.0 required `C++11`




## [v1.4](https://github.com/KoynovStas/uCRC_t/tree/v1.4) (2023-05-14)
[Full Changelog](https://github.com/KoynovStas/uCRC_t/compare/v1.3...v1.4)


### New
 - Pro: use cmake for build tests
 - add method `get_raw_crc` for first chunk of data


### Refactoring
 - add prefix `static` for `reflect` method
 - rename `get_final_crc` -> `get_end_crc`
 - update `STest` to ver 2.1


### Tests
 - add to CRC list `CRC-64/MS`
 - add to CRC list `CRC-64/REDIS`
 - add to CRC list `CRC-32/MEF`
 - add to CRC list `CRC-24/OS-9`
 - add to CRC list `CRC-21/CAN-FD`
 - add to CRC list `CRC-17/CAN-FD`
 - add to CRC list `CRC-16/M17`
 - add to CRC list `CRC-16/NRSC-5`
 - add to CRC list `CRC-8/HITAG`
 - add to CRC list `CRC-8/BLUETOOTH`
 - add to CRC list `CRC-8/NRSC-5`
 - add to CRC list `CRC-8/LTE`




## [v1.3](https://github.com/KoynovStas/uCRC_t/tree/v1.3) (2017-05-03)
[Full Changelog](https://github.com/KoynovStas/uCRC_t/compare/v1.2...v1.3)


### Added
- add method `get_check` - crc for ASCII string '123456789'


### Refactoring
 - Replacement `(const char *)` to `(const void *)` - for ease use methods: `get_crc` and `get_raw_crc`


### Tests
 - add to CRC list `CRC-64/GO-ISO`
 - add to CRC list `CRC-32/AUTOSAR`
 - add to CRC list `CRC-30/CDMA`
 - add to CRC list `CRC-24/LTE-B`
 - add to CRC list `CRC-24/LTE-A`
 - add to CRC list `CRC-24/INTERLAKEN`
 - add to CRC list `CRC-24/BLE`
 - add to CRC list `CRC-16/PROFIBUS`
 - add to CRC list `CRC-16/OPENSAFETY-B`
 - add to CRC list `CRC-16/OPENSAFETY-A`
 - add to CRC list `CRC-16/LJ1200`
 - add to CRC list `CRC-16/GSM`
 - add to CRC list `CRC-16/CMS`
 - add to CRC list `CRC-14/GSM`
 - add to CRC list `CRC-12/GSM`
 - add to CRC list `CRC-11/UMTS`
 - add to CRC list `CRC-10/GSM`
 - add to CRC list `CRC-8/SAE-J1850`
 - add to CRC list `CRC-8/OPENSAFETY`
 - add to CRC list `CRC-8/GSM-B`
 - add to CRC list `CRC-8/GSM-A`
 - add to CRC list `CRC-8/AUTOSAR`
 - add to CRC list `CRC-7/UMTS`
 - add to CRC list `CRC-6/GSM`
 - add to CRC list `CRC-4/INTERLAKEN`
 - add to CRC list `CRC-3/GSM`




## [v1.2](https://github.com/KoynovStas/uCRC_t/tree/v1.2) (2016-02-02)
[Full Changelog](https://github.com/KoynovStas/uCRC_t/compare/v1.1...v1.2)


### BugFix
 - In `get_crc()` for files must use `crc_init` NOT `init` [(fdcbe8d)](https://github.com/KoynovStas/uCRC_t/commit/fdcbe8dee781e01b6cc950900731c4fa725031bb)


### Tests
 - Add `test_crc_for_cunks`
 - Add `test_crc_no_file` for ALL CRC




## [v1.1](https://github.com/KoynovStas/uCRC_t/tree/v1.1) (2016-01-29)
[Full Changelog](https://github.com/KoynovStas/uCRC_t/compare/v1.0...v1.1)


### BugFix
 - BugFix: for Big Data **SEGFAULT** for `Bits` <= 8 [(ebc8f2c)](https://github.com/KoynovStas/uCRC_t/commit/ebc8f2c76ac4ff9a6e50b02c1203890ee3053c99)




## [v1.0](https://github.com/KoynovStas/uCRC_t/tree/v1.0) (2016-01-25)


crc_t.h
/*
 * crc_t.h
 *
 *
 * Copyright (c) 2015, Koynov Stas - skojnov@yandex.ru
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *  1 Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  2 Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  3 Neither the name of the <organization> nor the
 *    names of its contributors may be used to endorse or promote products
 *   derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

#ifndef CRC_T_H
#define CRC_T_H

#include <stdint.h>
#include <string>





class CRC_t
{

    public:

        explicit CRC_t(const std::string Name = "CRC-32",
                       uint8_t  Bits   = 32,
                       uint64_t Poly   = 0x04c11db7,
                       uint64_t Init   = 0xffffffff,
                       bool     RefIn  = true,
                       bool     RefOut = true,
                       uint64_t XorOut = 0xffffffff);


        explicit CRC_t(uint8_t  Bits,
                       uint64_t Poly,
                       uint64_t Init,
                       bool     RefIn,
                       bool     RefOut,
                       uint64_t XorOut);


        std::string name;


        // get param CRC
        uint8_t  get_bits()    const { return bits;   }
        uint64_t get_poly()    const { return poly;   }
        uint64_t get_init()    const { return init;   }
        uint64_t get_xor_out() const { return xor_out;}
        bool     get_ref_in()  const { return ref_in; }
        bool     get_ref_out() const { return ref_out;}

        uint64_t get_crc_init()const { return crc_init;} //crc_init = reflect(init, bits) if RefIn, else = init
        uint64_t get_top_bit() const { return top_bit; }
        uint64_t get_crc_mask()const { return crc_mask;}


        // set param CRC
        int  set_bits(uint8_t new_bits);
        void set_poly(uint64_t new_poly)       { poly    = new_poly; init_class();}
        void set_init(uint64_t new_init)       { init    = new_init; init_class();}
        void set_xor_out(uint64_t new_xor_out) { xor_out = new_xor_out;}
        void set_ref_in(bool new_ref_in)       { ref_in  = new_ref_in; init_class();}
        void set_ref_out(bool new_ref_out)     { ref_out = new_ref_out;}


        // Calculate methods
        uint64_t get_crc(const char* buf, size_t len)  const;
        int      get_crc(uint64_t *crc, const char *file_name)  const;

        // Calculate for chunks of data
        uint64_t get_raw_crc(uint64_t crc, const char* buf, size_t len) const; //for first byte crc = crc_init (must be)
        uint64_t get_final_crc(uint64_t raw_crc) const;



    private:

        uint8_t  bits;
        uint64_t poly;
        uint64_t init;
        bool     ref_in;
        bool     ref_out;
        uint64_t xor_out;

        uint8_t  shift;
        uint64_t crc_init;
        uint64_t top_bit;
        uint64_t crc_mask;
        uint64_t crc_table[256];


        uint64_t reflect(uint64_t data, uint8_t num_bits) const;
        void     init_crc_table();
        void     init_class();
};





#endif // CRC_T_H


crc_t.cpp
/*
 * crc_t.cpp
 *
 *
 * Copyright (c) 2015, Koynov Stas - skojnov@yandex.ru
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *  1 Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  2 Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  3 Neither the name of the <organization> nor the
 *    names of its contributors may be used to endorse or promote products
 *   derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

#include "crc_t.h"
#include <cstdio>




CRC_t::CRC_t(const std::string Name, uint8_t Bits, uint64_t Poly, uint64_t Init, bool RefIn, bool RefOut, uint64_t XorOut)
{
    name    = Name;
    bits    = Bits;
    poly    = Poly;
    init    = Init;
    ref_in  = RefIn;
    ref_out = RefOut;
    xor_out = XorOut;


    init_class();
}



CRC_t::CRC_t(uint8_t Bits, uint64_t Poly, uint64_t Init, bool RefIn, bool RefOut, uint64_t XorOut)
{
    bits    = Bits;
    poly    = Poly;
    init    = Init;
    ref_in  = RefIn;
    ref_out = RefOut;
    xor_out = XorOut;


    init_class();
}



int CRC_t::set_bits(uint8_t new_bits)
{
    if( (new_bits < 1) || (new_bits > 64) )
        return -1; //error


    bits = new_bits;
    init_class();


    return 0; //good job
}



uint64_t CRC_t::get_crc(const char *buf, size_t len) const
{
    uint64_t crc = get_raw_crc(crc_init, buf, len);

    return get_final_crc(crc);
}



int CRC_t::get_crc(uint64_t *crc, const char *file_name) const
{
    if( !file_name || !crc )
        return -1; //Bad param

    *crc = crc_init;

    char buf[4096];


    FILE *stream = fopen(file_name, "rb");
    if( stream == NULL )
        return -1; //Cant open file


    while( !feof(stream) )
    {
       size_t len = fread(buf, 1, sizeof(buf), stream);
       *crc = get_raw_crc(*crc, buf, len);
    }


    fclose(stream);


    *crc = get_final_crc(*crc);


    return 0; //good  job
}



uint64_t CRC_t::get_raw_crc(uint64_t crc, const char *buf, size_t len) const
{
    if(bits > 8)
    {
        if(ref_in)
            while (len--)
                crc = (crc >> 8) ^ crc_table[ (crc ^ *buf++) & 0xff ];
        else
            while (len--)
                crc = (crc << 8) ^ crc_table[ ((crc >> shift) & 0xff) ^ *buf++ ];
    }
    else
    {
        if (ref_in)
            while (len--)
                crc = crc_table[ crc ^ *buf++ ];
        else
            while (len--)
                crc = crc_table[ (crc << shift) ^ *buf++ ];
    }


    return crc;
}



uint64_t CRC_t::get_final_crc(uint64_t raw_crc) const
{
    if(ref_out^ref_in) raw_crc = reflect(raw_crc, bits);

    raw_crc ^= xor_out;
    raw_crc &= crc_mask; //for CRC not power 2

    return raw_crc;
}



uint64_t CRC_t::reflect(uint64_t data, uint8_t num_bits) const
{
    uint64_t reflection = 0;
    uint64_t one = 1;

    for ( size_t i = 0; i < num_bits; ++i, data >>= 1 )
    {
        if ( data & one )
        {
            reflection |= ( one << (num_bits - one - i) );
        }
    }

    return reflection;
}



void CRC_t::init_crc_table()
{
    int i;
    uint64_t crc;


    for(i = 0; i < 256; i++)
    {

        crc = 0;

        for(uint8_t mask = 0x80; mask; mask >>= 1)
        {

            if ( i & mask )
                crc ^= top_bit;


            if (crc & top_bit)
            {
                crc <<= 1;
                crc ^= poly;
            }
            else
                crc <<= 1;
        }

        crc &= crc_mask; //for CRC not power 2

        if(ref_in)
            crc_table[reflect(i, 8)] = reflect(crc, bits);
        else
            crc_table[i] = crc;
    }
}



void CRC_t::init_class()
{
    top_bit  = (uint64_t)1 << (bits - 1);
    crc_mask = ( (top_bit - 1) << 1) | 1;


    if(bits > 8)
        shift = (bits - 8);
    else
        shift = (8 - bits);


    if(ref_in)
        crc_init = reflect(init, bits);
    else
        crc_init = init;


    init_crc_table();
}


ucrc_t.h
/*
 * uCRC_t - is C++ class for calculation CRC any sizes 1-64 bits
 *
 *
 * version 1.4
 *
 *
 * BSD 3-Clause License
 *
 * Copyright (c) 2015, Koynov Stas - skojnov@yandex.ru
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 *
 * more details see https://github.com/KoynovStas/uCRC_t
 *
 */

#ifndef UCRC_T_H
#define UCRC_T_H

#include <cstdint>
#include <string>
#include <fstream>    // for std::ifstream
#include <ios>        // for std::ios_base, etc.





class uCRC_t
{
    public:
        explicit uCRC_t(std::string Name   = "CRC-32",
                        uint8_t     Bits   = 32,
                        uint64_t    Poly   = 0x04c11db7,
                        uint64_t    Init   = 0xffffffff,
                        bool        RefIn  = true,
                        bool        RefOut = true,
                        uint64_t    XorOut = 0xffffffff) noexcept;


        explicit uCRC_t(uint8_t  Bits,
                        uint64_t Poly,
                        uint64_t Init,
                        bool     RefIn,
                        bool     RefOut,
                        uint64_t XorOut) noexcept;


        std::string name;


        // get param CRC
        uint8_t  get_bits()    const noexcept { return bits;   }
        uint64_t get_poly()    const noexcept { return poly;   }
        uint64_t get_init()    const noexcept { return init;   }
        uint64_t get_xor_out() const noexcept { return xor_out;}
        bool     get_ref_in()  const noexcept { return ref_in; }
        bool     get_ref_out() const noexcept { return ref_out;}

        uint64_t get_top_bit() const noexcept { return top_bit; }
        uint64_t get_crc_mask()const noexcept { return crc_mask;}
        uint64_t get_crc_init()const noexcept { return crc_init;} //crc_init = reflect(init, bits) if RefIn, else = init
        uint64_t get_check()   const noexcept;                    //crc for ASCII string "123456789" (i.e. 313233... (hexadecimal)).


        // set param CRC
        int  set_bits(uint8_t new_value)     noexcept;
        void set_poly(uint64_t new_value)    noexcept { poly    = new_value; init_class();}
        void set_init(uint64_t new_value)    noexcept { init    = new_value; init_class();}
        void set_xor_out(uint64_t new_value) noexcept { xor_out = new_value;}
        void set_ref_in(bool new_value)      noexcept { ref_in  = new_value; init_class();}
        void set_ref_out(bool new_value)     noexcept { ref_out = new_value;}


        // Calculate methods
        uint64_t get_crc(const void* data, size_t len) const noexcept;
        int      get_crc(uint64_t &crc, const char* file_name) const noexcept;
        int      get_crc(uint64_t &crc, const char* file_name, void* buf, size_t size_buf) const noexcept;


        // Calculate for chunks of data
        uint64_t get_raw_crc(const void* data, size_t len) const noexcept;                   //get raw_crc for first chunk of data
        uint64_t get_raw_crc(const void* data, size_t len, uint64_t raw_crc) const noexcept; //get raw_crc for chunk of data
        uint64_t get_end_crc(uint64_t raw_crc) const noexcept;


    private:
        uint64_t poly;
        uint64_t init;
        uint64_t xor_out;
        uint64_t crc_init;  //crc_init = reflect(init, bits) if RefIn, else = init
        uint64_t top_bit;
        uint64_t crc_mask;
        uint64_t crc_table[256];

        uint8_t  bits;
        uint8_t  shift;
        bool     ref_in;
        bool     ref_out;


        static uint64_t reflect(uint64_t data, uint8_t num_bits) noexcept;
        void init_crc_table() noexcept;
        void init_class() noexcept;

        int  get_crc(uint64_t &crc, std::ifstream& ifs, void* buf, size_t size_buf) const noexcept;
};





inline uint64_t uCRC_t::get_check() const noexcept
{
    const uint8_t data[] = {0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39};

    return get_crc(data, sizeof(data));
}



inline int uCRC_t::set_bits(uint8_t new_value) noexcept
{
    if( (new_value < 1) || (new_value > 64) )
        return -1; //error

    bits = new_value;
    init_class();

    return 0; //good job
}



inline uint64_t uCRC_t::get_crc(const void* data, size_t len) const noexcept
{
    uint64_t crc = get_raw_crc(data, len, crc_init);

    return get_end_crc(crc);
}



inline int uCRC_t::get_crc(uint64_t& crc, const char* file_name) const noexcept
{
    char buf[4096];

    return get_crc(crc, file_name, buf, sizeof(buf));
}



inline uint64_t uCRC_t::get_raw_crc(const void* data, size_t len) const noexcept
{
    return get_raw_crc(data, len, crc_init);
}



inline uint64_t uCRC_t::get_end_crc(uint64_t raw_crc) const noexcept
{
    if(ref_out^ref_in) raw_crc = reflect(raw_crc, bits);

    raw_crc ^= xor_out;
    raw_crc &= crc_mask; //for CRC not power 2

    return raw_crc;
}





#endif // UCRC_T_H
